components/compositing/largest_contentful_paint.rs

```diff
@@ -0,0 +1,91 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
+
+use base::cross_process_instant::CrossProcessInstant;
+use compositing_traits::largest_contentful_paint_candidate::{
+    LCPCandidate, LargestContentfulPaint,
+};
+use fnv::{FnvBuildHasher, FnvHashMap};
+use webrender_api::{Epoch, PipelineId};
+
+/// Holds the [`LargestContentfulPaintCalculator`] for each pipeline.
+#[derive(Default)]
+pub(crate) struct LargestContentfulPaintDetector {
+    lcp_calculators: FnvHashMap<PipelineId, LargestContentfulPaintCalculator>,
+}
+
+impl LargestContentfulPaintDetector {
+    pub const fn new() -> Self {
+        Self {
+            lcp_calculators: FnvHashMap::with_hasher(FnvBuildHasher::new()),
+        }
+    }
+
+    fn ensure_lcp_calculator(
+        &mut self,
+        pipeline_id: PipelineId,
+    ) -> &mut LargestContentfulPaintCalculator {
+        self.lcp_calculators
+            .entry(pipeline_id)
+            .or_insert(LargestContentfulPaintCalculator {
+                lcp_candidates: Vec::new(),
+                latest_lcp: None,
+            })
```

Comment:
```suggestion
            .or_default()
```
nit: You can implement the `Default` trait for `LargestContentfulPaintCalculator` and then do this.